### PR TITLE
State: Migrate lib/shortcodes to Redux state

### LINKED ISSUE #7923: State: Migrate lib/shortcodes to Redux state
Related: #5046

The [`lib/shortcodes` Flux store](https://github.com/Automattic/wp-calypso/tree/master/client/lib/shortcodes) is a legacy store which uses deprecated data patterns. It should be updated to align with our [current data recommendations](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#current-recommendations). Specifically, this may include implementing all or some of the following Redux state behaviors:

- [ ] Reducer
- [ ] Action creators
- [ ] Selectors
- [ ] Query components

The implementation of each of these behaviors should be accompanied by tests.

When submitting a pull request, reference this issue and update the checklist above to reflect progress.

### PR DESCRIPTION
**Pull request by tyxla**
This PR creates the actions, reducer and selectors and query component for the shortcodes. Includes tests.

### Testing
From the terminal run `npm run test-client client/state/shortcodes/`

Fixes #7923

### CODE REVIEW

File: client/state/shortcodes/actions.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+export function fetchShortcode( siteId, shortcode ) {
+	return ( dispatch ) => {
+		dispatch( {
+			type: SHORTCODE_FETCH,
+			siteId,
+			shortcode
+		} );
+
+		return wpcom.undocumented().site( siteId ).shortcodes( {
+			shortcode: shortcode
```
Comment (aduth) [inline client/state/shortcodes/actions.js]: ES2015 includes a nice shorthand for creating objects where assigning a key to the same variable name:

```js
{ shortcode }
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015

File: client/state/shortcodes/actions.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		}, ( error, data ) => {
```
Comment (aduth) [inline client/state/shortcodes/actions.js]: Had you considered writing this as a Promise chain? You will need to make one minor revision to the [undocumented function](https://github.com/Automattic/wp-calypso/blob/e6eb37a38c8f5041e3cdf285bea03afc8ef989fc/client/lib/wpcom-undocumented/lib/site.js#L122) to `return` the `this.wpcom.req.get` call.

File: client/state/shortcodes/reducer.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ */
+import { createReducer } from 'state/utils';
+import { LOAD_STATUS } from './constants';
+import { shortcodesSchema } from './schema';
+import {
+	SHORTCODE_FETCH,
+	SHORTCODE_RECEIVE
+} from 'state/action-types';
+
+const siteShortcode = ( state = {}, { type, error, data } ) => {
```
Comment (aduth) [inline client/state/shortcodes/reducer.js]: Shortcode results can vary by site, so we'd need state to reflect this.

Also, rather than track "shortcode" and "shortcodes" separately, you might consider a single `items` object, which maps a site ID, shortcode pair to the received result.

It's fairly common in state to split out the request progress to its own reducer key. See [the page templates reducer](https://github.com/Automattic/wp-calypso/blob/e6eb37a38c8f5041e3cdf285bea03afc8ef989fc/client/state/page-templates/reducer.js#L18-L36) as an example. This could help eliminate the need for a constant and error handling within this reducer. To handle errors, you could create yet another reducer for tracking the error results of a shortcode request.

File: client/state/shortcodes/reducer.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	switch ( type ) {
+		case SHORTCODE_FETCH:
+			return { ...state, ...{ status: LOAD_STATUS.LOADING } };
```
Comment (aduth) [inline client/state/shortcodes/reducer.js]: This expression can be shortened to:

```js
return { ...state, status: LOAD_STATUS.LOADING };
```

Typically you'd only need the spread operator when adding properties from an existing object, but if you're creating a new one, you can just add the properties directly.

File: client/state/shortcodes/test/actions.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	SHORTCODE_RECEIVE
+} from 'state/action-types';
+import { fetchShortcode } from '../actions';
+import { useSandbox } from 'test/helpers/use-sinon';
+import wpcom from 'lib/wp';
+
+describe( 'actions', () => {
+	let sandbox, spy;
+
+	useSandbox( newSandbox => {
```
Comment (aduth) [inline client/state/shortcodes/test/actions.js]: I think this could be simplified to:

```js
let spy;
useSandbox( ( sandbox ) => spy = sandbox.spy() );
```

File: client/state/shortcodes/test/actions.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+						shortcode
+					} )
+					.reply( 400, {
+						error: 'The requested shortcode does not exist.'
+					}, {
+						'Content-Type': 'application/json'
+					} );
+			} );
+
+			after( () => {
```
Comment (aduth) [inline client/state/shortcodes/test/actions.js]: The [`useNock` test helper](https://github.com/Automattic/wp-calypso/tree/master/test/test/helpers/use-nock) can help avoid the need for nock cleanup. See also #7875 where I update existing usage.

File: client/state/shortcodes/actions.js (hunk 1 above)
Comment (tyxla) [inline client/state/shortcodes/actions.js]: Totally missed that - done.

File: client/state/shortcodes/actions.js (hunk 2 above)
Comment (tyxla) [inline client/state/shortcodes/actions.js]: Sure, good idea - done.

File: client/state/shortcodes/test/actions.js (hunk 5 above)
Comment (tyxla) [inline client/state/shortcodes/test/actions.js]: Sure, done.

File: client/state/shortcodes/test/actions.js (hunk 6 above)
Comment (tyxla) [inline client/state/shortcodes/test/actions.js]: Nice! Addressed in all action tests.

**Comment by tyxla** [comment]
@omarjackman we'll also need to handle the `RECEIVE_MEDIA_ITEM` and `RECEIVE_MEDIA_ITEMS` actions in the reducer.

**Comment by aduth** [comment]
>@omarjackman we'll also need to handle the `RECEIVE_MEDIA_ITEM` and `RECEIVE_MEDIA_ITEMS` actions in the reducer.

Ooh, that's a little tricky... since media is still Flux-based.

**Comment by tyxla** [comment]
> Ooh, that's a little tricky... since media is still Flux-based.

Yeah, indeed - we discussed that with Omar at the GM. I'm just throwing that here so we don't accidentally forget about it.

**Comment by tyxla** [comment]
I've made some significant changes and improvements - the PR is now ready to review. The next step once this lands is #8857 which takes advantage of the shortcodes redux state. 

If you've checked out this branch previously, make sure to `git pull --rebase`.

/cc @aduth @ockham

File: client/state/action-types.js [hunk 7]
```diff
@@ -361,7 +361,12 @@ export const SELECTED_SITE_SET = 'SELECTED_SITE_SET';
 export const SERIALIZE = 'SERIALIZE';
 export const SERVER_DESERIALIZE = 'SERVER_DESERIALIZE';
 export const SEO_TITLE_SET = 'SEO_TITLE_SET';
+export const SHORTCODE_RECEIVE = 'SHORTCODE_RECEIVE';
+export const SHORTCODE_REQUEST = 'SHORTCODE_REQUEST';
+export const SHORTCODE_REQUEST_FAILURE = 'SHORTCODE_REQUEST_FAILURE';
+export const SHORTCODE_REQUEST_SUCCESS = 'SHORTCODE_REQUEST_SUCCESS';
 export const SIGNUP_COMPLETE_RESET = 'SIGNUP_COMPLETE_RESET';
+export const SIGNUP_DEPENDENCY_STORE_RESET = 'SIGNUP_DEPENDENCY_STORE_RESET';
```
Comment (aduth) [inline client/state/action-types.js]: Why was this line added? Bad rebase?

File: client/state/shortcodes/actions.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+				type: SHORTCODE_REQUEST_SUCCESS,
+				siteId,
+				shortcode
+			} );
+
+			dispatch( {
+				type: SHORTCODE_RECEIVE,
+				siteId,
+				shortcode,
+				data: {
```
Comment (aduth) [inline client/state/shortcodes/actions.js]: Couple thoughts:

- I know it's how it was previously, but is there value in changing the structure to reassign `result` as `body` ? Would simplify here to simply dispatch `data`.
- If we were to continue transforming the structure, do you think this should be done in the reducer instead?

File: client/state/shortcodes/reducer.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import {
+	SHORTCODE_RECEIVE,
+	SHORTCODE_REQUEST,
+	SHORTCODE_REQUEST_FAILURE,
+	SHORTCODE_REQUEST_SUCCESS
+} from 'state/action-types';
+
+const requestingShortcode = ( state, siteId, shortcode, requesting ) => {
+	const siteShortcodes = has( state, siteId ) ? state[ siteId ] : {};
+	return {
```
Comment (aduth) [inline client/state/shortcodes/reducer.js]: [Lodash `_.merge`](https://lodash.com/docs/4.16.4#merge) can be nice for this type of deep assign:

```js
return merge( {}, state, {
	[ siteId ]: {
		[ shortcode ]: requesting
	}
} );
```

Only time it gets you in trouble is when properties on a nested item need to be unset.

As is though, is an error thrown if you try to use the spread where the existing value would be `undefined`? Thinking:

```js
return {
	...state,
	[ siteId ]: {
		...state[ siteId ],
		[ shortcode ]: requesting
	}
};
```

File: client/state/shortcodes/schema.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+export const siteShortcodesSchema = {
+	type: 'object',
+	patternProperties: {
+		'^.+$': { ...siteShortcodeSchema }
+	}
+};
+
+export const shortcodesSchema = {
```
Comment (aduth) [inline client/state/shortcodes/schema.js]: Why three schemas if only one is used?

File: client/state/shortcodes/selectors.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ * Returns true if currently requesting that shortcode for the specified site ID, or
+ * false otherwise.
+ *
+ * @param  {Object}  state     Global state tree
+ * @param  {Number}  siteId    Site ID
+ * @param  {String}  shortcode Shortcode
+ * @return {Boolean}           Whether that shortcode is being requested
+ */
+export const isRequestingShortcode = ( state, siteId, shortcode ) => {
+	if ( ! has( state, [ 'shortcodes', 'requesting', siteId, shortcode ] ) ) {
```
Comment (aduth) [inline client/state/shortcodes/selectors.js]: You can be certain that at least the `state.shortcodes.requesting` part of the path will always exist, so it could save a few cycles in Lodash to avoid checking whether they're accessible.

You could also use the third argument to [`_.get`](https://lodash.com/docs/4.16.4#get) to simplify the fallback:

```js
return get( state.shortcodes.requesting, [ siteId, shortcode ], false );
```

Or:

```js
return !! get( state.shortcodes.requesting, [ siteId, shortcode ] );
```

File: client/state/shortcodes/selectors.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * @param  {Number} siteId     Site ID
+ * @param  {String} shortcode  Shortcode
+ * @return {Object}            Shortcode data
+ */
+export const getShortcode = ( state, siteId, shortcode ) => {
+	if ( ! has( state, [ 'shortcodes', 'items', siteId, shortcode ] ) ) {
+		return;
+	}
+
+	return state.shortcodes.items[ siteId ][ shortcode ];
```
Comment (aduth) [inline client/state/shortcodes/selectors.js]: Similar to above:

```js
return get( state.shortcodes.items, [ siteId, shortcode ] );
```

File: client/state/shortcodes/actions.js (hunk 8 above)
Comment (tyxla) [inline client/state/shortcodes/actions.js]: Actually, having that as a `result` looks perfectly fine to me. This simplifies everything, so I've changed it to not convert `result` to `body`.

File: client/state/shortcodes/reducer.js (hunk 9 above)
Comment (tyxla) [inline client/state/shortcodes/reducer.js]: `merge()` works great for that purpose indeed, thanks (we're not unsetting anyway)! And `...state[ siteId ]` also works fine, as long as state is not `null` for some reason (which it should not be). I've updated the PR with both of these, it looks very simplified now.

File: client/state/shortcodes/schema.js (hunk 10 above)
Comment (tyxla) [inline client/state/shortcodes/schema.js]: That remained from the previous version of the reducer. Merged into one schema now.

File: client/state/shortcodes/selectors.js (hunk 11 above)
Comment (tyxla) [inline client/state/shortcodes/selectors.js]: Good point :+1:

**Comment by tyxla** [comment]
Thanks for the review @aduth! I've cleaned up a little after your suggestions, feel free to have another look.

File: client/components/data/query-shortcode/README.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+```jsx
+import React from 'react';
+import QueryShortcode from 'components/data/query-shortcode';
+import MySiteShortcodeItemDetail from './item-detail';
+
+export default function MySiteShortcodeItem( { shortcode } ) {
+	return (
+		<div>
+			<QueryShortcode siteId={ 12345678 } />
```
Comment (aduth) [inline client/components/data/query-shortcode/README.md]: Example should include `shortcode` prop.

File: client/components/data/query-shortcode/index.jsx [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+		props.fetchShortcode( props.siteId, props.shortcode );
+	}
+
+	render() {
+		return null;
+	}
+}
+
+export default connect(
+	( state, ownProps ) => {
```
Comment (aduth) [inline client/components/data/query-shortcode/index.jsx]: Curious how you feel about a more compact variant:

```js
( state, { siteId, shortcode } ) => ( {
	requestingShortcode: isRequestingShortcode( state, siteId, shortcode )
} )
```

File: client/state/shortcodes/reducer.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Returns the updated requests state after an action has been dispatched. The
+ * state maps site ID keys to the object with the site's shortcodes. Each shortcode
+ * is true if it's being currently requested, false otherwise.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export const requesting = createReducer( {}, {
```
Comment (aduth) [inline client/state/shortcodes/reducer.js]: This is fine, but you could avoid some of the boilerplate of passing arguments by having the function return the reducer handler:

```js
const createRequestingReducer = ( requesting ) => {
	return ( state, { siteId, shortcode } ) => {
		return merge( {}, state, {
			[ siteId ]: {
				[ shortcode ]: requesting
			}
		} );
	};
};

export const requesting = createReducer( {}, {
	[ SHORTCODE_REQUEST ]: createRequestingReducer( true ),
	[ SHORTCODE_REQUEST_FAILURE ]: createRequestingReducer( false ),
	[ SHORTCODE_REQUEST_SUCCESS ]: createRequestingReducer( false ),
} );
```

File: client/state/shortcodes/schema.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+	type: 'object',
+	patternProperties: {
+		'^\\d+$': {
+			type: 'object',
+			patternProperties: {
+				'^.+$': {
+					type: 'object',
+					properties: {
+						result: { type: 'string' },
+						shortcode: { type: 'string' },
```
Comment (aduth) [inline client/state/shortcodes/schema.js]: I don't feel strongly one way or the other, and we don't have a strong history of modifying response values, but since `shortcode` is redundant with the key of the object, we could `omit` (or conversely `pick`) this from the reducer state.

(related to https://github.com/Automattic/wp-calypso/pull/8116#discussion_r84277962)

File: client/components/data/query-shortcode/README.md (hunk 13 above)
Comment (tyxla) [inline client/components/data/query-shortcode/README.md]: Good catch 👍

File: client/components/data/query-shortcode/index.jsx (hunk 14 above)
Comment (tyxla) [inline client/components/data/query-shortcode/index.jsx]: Even better, thanks!

File: client/state/shortcodes/reducer.js (hunk 15 above)
Comment (tyxla) [inline client/state/shortcodes/reducer.js]: Didn't occur to me, thanks!

**Comment by tyxla** [comment]
I feel this is ready to go now. Any final thoughts @aduth?
